### PR TITLE
Unsteady Driver Fix + Piston Unsteady

### PR DESCRIPTION
**Pull request by sean-engelstad**
- Mistake in the unsteady derivative tests involved using a steady scenario instead of unsteady. This was fixed and through the following changes the unsteady solve_forward() and solve_adjoint() are compatible.
- Added piston theory unsteady test however it does not pass yet and is still in progress. Computes dwdt based on saved piston theory ScenarioData now and therefore pressures are modified due to dw/dt.
- Jump condition added to unsteady funtofem driver solve_forward() (`FUNtoFEMnlbgs`) so that state variables are transferred between time steps. They did not interact before and therefore did not pass the new unsteady derivative tests.
- Simplified the `TacsIntegrationSettings` with the useful @dataclass decorator which builds its own constructor method in python based on the provided data attributes.
- Easier construction of piston theory interface object using @dataclass for new classes `PistonTheoryGrid` and `PistonTheoryFlow` with their own @classmethods
- Aerothermal and Aerothermoelastic unsteady tests do not pass to machine precision yet with TACSinterface.

### CODE REVIEW

**Comment by bburke38** [comment]
Why is `thermal_scheme` no longer returning self in `transfer_settings.py`?

**Comment by sean-engelstad** [comment]
Oops let's make thermal_scheme return self again.

**Comment by sean-engelstad** [comment]
Also, I had tested the piston steady unittest several times and had gotten stuck. I got the same answer replacing the old piston theory interface source code over the current and running this test. So really points to the difference in the new piston theory test which uses aoa and thickness. Will have to resolve this

**Comment by sean-engelstad** [comment]
Another comment so I don't forget. There is a new piston unsteady test which fails with 0 derivative compared to 0 complex step (so indeterminate). The reason is piston theory computes zero pressure load for zero aero disps which I think is wrong. I believe the CD_mat finite difference matrix is wrong and therefore results in no aero loads for initial zero aero disps. For that reason, the piston theory unsteady tests has zero state variables all throughout the test. In fun3d and other flow solvers, the flow solver starts the nonzero state variables with nonzero aero loads for zero aero disps.

**Comment by bburke38** [comment]
Okay, I'll have to look at piston theory itself to confirm, I'm not sure if that is correct. Also, why are the fail checks in the nlbgs driver commented out?

**Comment by bburke38** [comment]
Also, it still looks like thermal_scheme isn't returning self.

**Comment by sean-engelstad** [comment]
It's not ready yet. I am still working on it. Just had to push what I was working on so I can access it in multiple places

File: pyfuntofem/driver/funtofem_nlbgs_driver.py [hunk 1]
```diff
@@ -402,57 +326,26 @@ def _solve_unsteady_adjoint(self, scenario):
...
 
-            self.solvers.flow.set_states(scenario, self.model.bodies, step)
+            self.solvers.flow.set_states(scenario, self.model.bodies, rev_step)
             # Due to the staggering, we linearize the transfer about t_s^(n-1)
-            self.solvers.structural.set_states(scenario, self.model.bodies, step - 1)
+            self.solvers.structural.set_states(scenario, self.model.bodies, rev_step)
 
             for body in self.model.bodies:
                 body.transfer_disps_adjoint(scenario)
-                # body.transfer_loads_adjoint(scenario)
```
Comment (sean-engelstad) [inline pyfuntofem/driver/funtofem_nlbgs_driver.py:421]: Are these adjoints sensitive to the state variables of the transfer scheme? Maybe some of them are not and others are, like the load adjoint. Double check this

File: pyfuntofem/driver/funtofem_nlbgs_driver.py [hunk 2]
```diff
@@ -461,161 +354,43 @@ def _solve_unsteady_adjoint(self, scenario):
...
-            #                        # funtofem: dQdftA^T * psi_Q = dTdts * psi_Q
-            #                        psi_Q_r = np.zeros(body.aero_nnodes, dtype=TransferScheme.dtype)
-            #                        body.thermal_transfer.applydQdqATrans(
-            #                            body.psi_T_S[:, func].copy(order="C"), psi_Q_r
-            #                        )
-            #                        body.dQdfta[:, func] = psi_Q_r
-
             for body in self.model.bodies:
                 body.transfer_loads_adjoint(scenario)
                 body.transfer_heat_flux_adjoint(scenario)
```
Comment (sean-engelstad) [inline pyfuntofem/driver/funtofem_nlbgs_driver.py:359]: Do we need to set states in before the load adjoint or heat flux adjoint?

**Pull request closed without merging**